internal/operators/cnv/cnv_operator.go

```diff
@@ -0,0 +1,139 @@
+package cnv
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/openshift/assisted-service/internal/common"
+	"github.com/openshift/assisted-service/internal/hardware"
+	"github.com/openshift/assisted-service/internal/operators/api"
+	models "github.com/openshift/assisted-service/models"
+	"github.com/sirupsen/logrus"
+)
+
+type operator struct {
+	log logrus.FieldLogger
+}
+
+// NewCNVOperator creates new instance of a Container Native Virtualization installation plugin
+func NewCNVOperator(log logrus.FieldLogger) *operator {
+	return &operator{
+		log: log,
+	}
+}
+
+// GetType reports type of an operator this Operator manages
+func (o *operator) GetType() models.OperatorType {
+	return models.OperatorTypeCnv
+}
+
+// GetDependencies provides a list of dependencies of the Operator
+func (o *operator) GetDependencies() []models.OperatorType {
+	return []models.OperatorType{models.OperatorTypeLso}
+}
+
+// GetClusterValidationID returns cluster validation ID for the Operator
+func (o *operator) GetClusterValidationID() string {
+	return string(models.ClusterValidationIDCnvRequirementsSatisfied)
+}
+
+// GetHostValidationID returns host validation ID for the Operator
+func (o *operator) GetHostValidationID() string {
+	return string(models.HostValidationIDCnvRequirementsSatisfied)
+}
+
+// ValidateCluster always return "valid" result
+func (o *operator) ValidateCluster(ctx context.Context, cluster *common.Cluster) (api.ValidationResult, error) {
+	var insufficientHosts []string
+	var status string
+	for _, host := range cluster.Hosts {
+		result, err := o.ValidateHost(ctx, cluster, host)
+		if err != nil {
+			return api.ValidationResult{Status: api.Failure, ValidationId: o.GetClusterValidationID()}, err
+		}
+		if result.Status != api.Success {
+			insufficientHosts = append(insufficientHosts, fmt.Sprint("Insufficient resources on host with host ID ", *host.ID, " to deploy CNV"))
+		}
+	}
+	if len(insufficientHosts) > 0 {
+		for _, hostStatus := range insufficientHosts {
+			status = status + hostStatus + ".\n"
+		}
+		err := setOperatorStatus(&cluster.Cluster, status)
+		if err != nil {
+			o.log.Error("Failed to set Operator status ", err)
+		} else {
+			o.log.Info("Setting Operator Status ", status)
+		}
+		return api.ValidationResult{Status: api.Failure, ValidationId: o.GetClusterValidationID(), Reasons: []string{status}}, err
+	}
+	return api.ValidationResult{Status: api.Success, ValidationId: o.GetClusterValidationID()}, nil
+}
+
+// ValidateHost always return "valid" result
+func (o *operator) ValidateHost(ctx context.Context, cluster *common.Cluster, host *models.Host) (api.ValidationResult, error) {
```

Comment:
I think that the validateHost method should return also the validation failure, for example: not enough memory for worker: expected: X, exists: Y - it will be more informative for the user and we should add it to the logs for debugging

Reply:
As above. It is the [interface](https://github.com/openshift/assisted-service/blob/master/internal/operators/manager.go#L23) defined by @jakub-dzon Do you wante me to update it?

Reply:
The `ValidationResult` struct has a `Reasons` field, where you can pass human-readable validation failure info.

Reply:
Yes, what @jakub-dzon said, don't need to add another parameter to the response - my mistake. Just update the validation info inside this struct.